### PR TITLE
Feat(eos_designs): Inject default VRF policy with a match-all statement when missing

### PR DESCRIPTION
**Pull request by gmuloc**
## Change Summary

In the current implementation, when there is no policy defined under `wan_virtual_topology.vrfs` for VRF default. AVD will raise an error.

This PR replaces this behavior with auto configuring a policy with one default statement (`default_virtual_topology`) matching all local path_groups in the load balance policy.

Of course, because it is VRF default, the Control plane special virtual topology is still injected.

## Component(s) name

`arista.avd.eos_designs`

## Proposed changes

When the VRF default is missing from `wan_virtual_topology.vrfs`, AVD creates a policy called `DEFAULT-AVT-POLICY` that is configured with one default statement to match all traffic using all locally available path-groups

## How to test
molecule updated to test it in AutoVPN use case.

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/AUTOVPN_TESTS.yml [hunk 1]
```diff
@@ -100,8 +100,7 @@ tenants:
 
 wan_virtual_topologies:
   vrfs:
-    - name: default
-      policy: DEFAULT-AVT-POLICY
+    # Not configuring default VRF policy on purpose to test auto-generation
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/AUTOVPN_TESTS.yml]: Why are these changes causing the IT stuff to get removed?

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/autovpn-rr1.cfg [hunk 2]
```diff
@@ -42,10 +41,10 @@ router path-selection
       path-group INET
    !
    policy DEFAULT-AVT-POLICY-WITH-CP
+      default-match
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/autovpn-rr1.cfg]: The order on EOS is opposite I believe? default-match being last.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/AUTOVPN_TESTS.yml (hunk 1 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/AUTOVPN_TESTS.yml]: Ah because you are adding CUSTOM-DEFAULT-AVD-POLICY but not referring to it anywhere. We are loosing a lot of test coverage with this change.

Comment (gmuloc) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/AUTOVPN_TESTS.yml]: this is covered in CV PATHFINDER

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/autovpn-rr1.cfg (hunk 2 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/autovpn-rr1.cfg]: not for router path-selction policies

```
router path-selection
   tcp mss ceiling ipv4 ingress
   !
   policy BLAH
      default-match
         load-balance BLAH
      !
      10 application-profile TITI
         load-balance TITI

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/AUTOVPN_TESTS.yml (hunk 1 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/AUTOVPN_TESTS.yml]: otherwise I need two new examples in eos_designs just for this

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py [hunk 3]
```diff
@@ -316,14 +316,24 @@ def _wan_load_balance_policies(self) -> list:
                     context_keys=["name"],
                 )
 
-            default_virtual_topology = get(policy, "default_virtual_topology", required=True)
+            default_virtual_topology = get(
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py:322]: Something seems wrong about the order here. You error out suggesting the drop_unmatched _above_ the line where you check for drop_unmatched.

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py [hunk 4]
```diff
@@ -316,14 +316,24 @@ def _wan_load_balance_policies(self) -> list:
...
+                ),
+            )
             if not get(default_virtual_topology, "drop_unmatched", default=False):
                 name = get(default_virtual_topology, "name", default=f"{policy['name']}-DEFAULT")
                 context_path = f"wan_virtual_topologies.policies[{policy['name']}].default_virtual_topology"
                 append_if_not_duplicate(
                     list_of_dicts=wan_load_balance_policies,
                     primary_key="name",
-                    new_dict=self._generate_wan_load_balance_policy(f"LB-{name}", default_virtual_topology, context_path),
+                    new_dict=self._generate_wan_load_balance_policy(
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py]: Why not just do the _default_all check in the function you call instead of here where it seems a bit "clunky"?

**Review by ClausHolbechArista — COMMENTED** [review]
The logic becomes increasingly hard to follow. More comments in the code might help.

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py (hunk 3 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py:322]: so not really - line 319 is checking for the presence of the key that must be there
the second line is checking that either `drop_unmatched` is set (and then we render nothing) or then we render path-groups.

This is to force explicit configuration of `defaut_virtual_topology` every time as we discussed - if you want to drop unmatched then you need to do it explicitly rather than having it done silently for you (I need to check if the dict is required in schema)

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py (hunk 4 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py]: hmm the mechanism using input was already used some other place for control_plane LB policy and here I am reusing this but I can just use the key instead and change the behavior. Let me refactor to make it less clunky

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py (hunk 3 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py:322]: The error message on line 319 is that the default_virtual_topology must be specified. This will error out no matter what since you have required=True. So setting `drop_unmatched` to True will not stop that.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/AUTOVPN_TESTS.yml (hunk 1 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/AUTOVPN_TESTS.yml]: You could just add another VRF with the old policy, so we keep all the checks also for autovpn case.

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py (hunk 3 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py:322]: yes thats the goal
1. you must define `default_virtual_topology` with either `drop_unmatched` or something else - line 319 checks that it is defined if not it errors out
2. either you have `drop_unmatched` and nothing is rendered or we render the rest

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py:322]: For someone getting this error, this last line will only confuse them. I think this org_key should just be the full variable path to the missing key. But it would be simpler to just require it in the schema and just use `['default_virtual_topology']` here.

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py [hunk 5]
```diff
@@ -316,14 +316,24 @@ def _wan_load_balance_policies(self) -> list:
...
-            default_virtual_topology = get(policy, "default_virtual_topology", required=True)
+            default_virtual_topology = get(
+                policy,
+                "default_virtual_topology",
+                required=True,
+                org_key=(
+                    f"A 'default_virtual_topology must be defined for policy {policy['name']}. "
+                    "It is possible to disable default-match by setting 'drop_unmatched' to 'true'."
+                ),
+            )
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py:324]: check for policy keyword line 329 and raise if not present

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py [hunk 6]
```diff
@@ -391,25 +401,30 @@ def _filtered_wan_policies(self) -> list:
...
+                vrf_default_policy,
+                required=True,
+                custom_error_msg=(
+                    f"The policy {vrf_default_policy} defined for vrf default under 'wan_virtual_topologies.vrfs' "
+                    "is not defined under 'wan_virtual_topologies.policies'."
+                ),
+            ).copy()
+        else:
+            # Injecting a hidden key _default_all, used when generating the relevant Load Balance Policy
+            default_policy = {"name": "DEFAULT-AVT-POLICY", "default_virtual_topology": {"_default_all": True}}
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py]: use all path-groups here and in control-plane to avoid extra flags

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/AUTOVPN_TESTS.yml (hunk 1 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/AUTOVPN_TESTS.yml]: adding new hosts instead for thsi specifically

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py [hunk 7]
```diff
@@ -265,6 +254,17 @@ def _generate_wan_load_balance_policy(self, name: str, input_dict: dict, context
 
         return wan_load_balance_policy
 
+    @cached_property
+    def _local_path_groups_connected_to_pathfinder(self) -> list:
+        """
+        Return list of local path_groups connected to pathfinder
+        """
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/utils.py]:
```suggestion
    def _local_path_groups_connected_to_pathfinder(self) -> list:
        """
        Return list of names of local path_groups connected to pathfinder
        """
```

**Review by ClausHolbechArista — APPROVED** [review]

**Review by carlbuchmann — APPROVED** [review]
LGTM, works as expected in test topology
